building-protected-and-paywall.mdx

```diff
@@ -0,0 +1,969 @@
+  // ...
+
+  public async setSession(databaseSession: DatabaseSession): Promise<void> {
+    await this.controller.db.session.create({
+      id: databaseSession.id,
+      user_id: databaseSession.userId,
+      expires_at: databaseSession.expiresAt,
+      ...databaseSession.attributes
+    });
+  }
+}
+```
+
+The code above defines a `setSession` method that uses Xata SDK methods to insert a record in session table in your database. Define a method to update a session by appending the following code:
+
+```tsx
+// File: src/lucia/xata.ts
+
+// ...
+
+export class XataAdapter implements Adapter {
+  // ...
+
+  public async updateSessionExpiration(sessionId: string, expiresAt: Date): Promise<void> {
+    await this.controller.sql`UPDATE "session" SET expires_at=${expiresAt} WHERE id=${sessionId}`;
+  }
+}
+```
+
+The code above defines an `updateSessionExpiration` method that uses Xata SQL query to update the `expires_at` attribute in a particular session in your database. Define a method to delete all expired sessions by appending the following code:
+
+```tsx
+// File: src/lucia/xata.ts
+
+// ...
+
+export class XataAdapter implements Adapter {
+  // ...
+
+  public async deleteExpiredSessions(): Promise<void> {
+    await this.controller.sql`DELETE FROM "session" WHERE expires_at <=${new Date()}`;
+  }
+}
+```
+
+The code above defines a `deleteExpiredSessions` method that uses Xata SQL query to delete all the sessions whose `expires_at` attribute have expired. Define a method to obtain a session by appending the following code:
+
+```tsx
+// File: src/lucia/xata.ts
+
+// ...
+
+export class XataAdapter implements Adapter {
+  // ...
+
+  private async getSession(session_id: string): Promise<DatabaseSession | null> {
+    const records = await this.controller.db.session.filter({ session_id }).getFirst();
+    return transformIntoDatabaseSession(records);
+  }
+}
+```
+
+The code above defines a `getSession` method that uses Xata SDK method to retrieve a single matching record for a given session from your database. Define a method to obtain a user associated with a session by appending the following code:
+
+```tsx
+// File: src/lucia/xata.ts
+
+// ...
+
+export class XataAdapter implements Adapter {
+  // ...
+
+  private async getUserFromSessionId(session_id: string): Promise<DatabaseUser | null> {
+    const theSession = await this.controller.db.session.filter({ session_id }).getFirst();
+    if (theSession) {
+      const { user_id } = theSession;
+      const getUser = await this.controller.db.user.filter({ user_id }).getFirst();
+      if (getUser) return transformIntoDatabaseUser(getUser);
+    }
+    return null;
+  }
+}
+```
+
+The code above defines a `getUserFromSessionId` method that uses Xata SDK method to retrieve a single matching record for a given session from your database. If a session is found, it further queries to obtain the user associated with the session. Otherwise, it returns `null`.
```

Comment:
```suggestion
The code above defines a `getUserFromSessionId` method that uses Xata SDK methods to retrieve a single matching record for a given session from your database. If a session is found, it further queries to obtain the user associated with the session. Otherwise, it returns `null`.
```